torch/csrc/jit/passes/graph_fuser.cpp

```diff
@@ -382,6 +384,94 @@ struct GraphFuser {
     return ++consumer->reverseIterator();
   }
 
+  // This is a hack to improve the fusion in case the graph looks like this:
+  //
+  // %1 = f(...)
+  // %2 = g(%1)
+  // %3 = h(%1)
+  // %4 = l(%3)
+  // return %4, %2
+  //
+  // where f, g, h, l are fusable operations
+  //
+  // If you follow the algorithm, you'll see that we will end up with a group
+  // that fuses 3 and 4, and another one that fuses 2 and 1. This function will
+  // connect them into a single group.
+  void mergeNeighbouringFusionGroups() {
```

Comment:
How does this handle when the fusion groups have to be separate for topological reasons? It is not obvious from the code:

```
a = fusion0(b);
c = black_box(a);
d = fusion1(a,c);
```

Reply:
hmm ok, good point. I'll think about sth better 😕

Reply:
Ok, I think it should be relatively easy to check if it's safe to merge them: follow the def->use chains for all nodes topologically between them, and allow their fusion only if there is no other path from producer to consumer.

Or do you have a better idea for a fix for that graph I mentioned in the comment?

Reply:
If you want to add that then you can also just put it in the normal fuser path :) It ignored FusionGroup<->FusionGroup fusion because it was annoying to do, but you wrote that code already so if you also right the safety check code, you should be good to go.

A simpler but more conservative safety check: all uses of producer outputs that are not from the consumer happen topologically after the consumer. This check is already in the code.

Reply:
I don't think that the main path ignores FusionGroup <-> FusionGroup - it's just impossible to trigger this condition. We always iterate from nodes of higher topological index, and fusion groups have the index equal to max of nodes they fuse. This means that once we finish fusing things into group X, all other nodes considered as consumers will have topological numbers strictly less than X, and you will never find X in their inputs, so there's no way to even consider this possibility.

The only way we could do it would be to run the fusing pass multiple times, until it makes no changes to the graph. This is an alternative that I thought about, but decided that this is simpler. However, as you said, now that the code is there it shouldn't be too hard, so I can do it.

Yeah, that works too. Checking the more relaxed condition shouldn't be too computationally expensive, because you limit yourself to this particular topological range (I think it's n^2 worst case, but that will almost never happen in real graphs). But this could be also relaxed in the main path, so we can just consider doing it later.

Reply:
I see. I think the simplest change would be:
* iterate until no new fusions are made (a common approach to optimizations)
* support fusiongroup-fusiongroup using the conservative check.

Reply:
Yup. That's exactly what I was planning to do now 👍

Reply:
Done. I only had to do tiny modifications to the code to get this working properly.